napalm/base/utils/string_parsers.py

```diff
@@ -133,3 +134,14 @@ def convert_uptime_string_seconds(uptime: str) -> int:
         raise Exception("Unrecognized uptime string:{}".format(uptime))
 
     return uptime_seconds
+
+
+def parse_fixed_width(text: str, *fields: int) -> List[Tuple[Any, ...]]:
+    len = sum(fields)
+    fmtstring = " ".join(f"{fw}s" for fw in fields)
+    unpack = struct.Struct(fmtstring).unpack_from
+
+    def parse(line: str) -> Tuple[Any, ...]:
+        return tuple(s.decode() for s in unpack(line.ljust(len).encode()))
```

Comment:
should I explicitly encode as "ascii" here just to make sure we don't end up with weird output from the guy who puts UTF-8 into VRF names somehow?

Reply:
I think I would just leave it as utf-8 i.e. that is probably the safer option (in general).

But let me know if I am missing something here (i.e. it is possible I am missing the specific issue you are concerned about).

Reply:
... I am not advocating for this.  It is possible to use unicode emojis in switch configs.

```python
In [1]: from napalm.eos import EOSDriver

In [2]: d = EOSDriver("172.20.20.5", username="admin", password="admin")

In [3]: d.open()

In [4]: d.load_merge_candidate(config=["interface Ethernet1", "description delicious 🧁"])

In [5]: d.commit_config()

In [6]: d.get_interfaces()["Ethernet1"]
Out[6]:
{'is_up': True,
 'is_enabled': True,
 'description': 'delicious 🧁',
 'last_flapped': 1684246566.1269095,
 'mtu': 1500,
 'speed': 1000.0,
 'mac_address': 'AA:C1:AB:C2:FF:EB'}

In [7]:
```

```console
r1#show interface Et1
Ethernet1 is up, line protocol is up (connected)
  Hardware is Ethernet, address is aac1.abc2.ffeb (bia aac1.abc2.ffeb)
  Description: delicious <U+1F9C1>
```

Attempting to use `struct` to handle variable width UTF-8 strings is no good:

```python
In [5]: parse_fixed_width("12345689🧁1234567890123456", 10, 10, 10)
---------------------------------------------------------------------------
UnicodeDecodeError                        Traceback (most recent call last)
Cell In[5], line 1
----> 1 parse_fixed_width("12345689🧁1234567890123456", 10, 10, 10)

File ~/projects/napalm/napalm/base/utils/string_parsers.py:147, in parse_fixed_width(text, *fields)
    144 def parse(line: str) -> Tuple[str, ...]:
    145     return tuple([str(s.decode()) for s in unpack(line.ljust(len).encode())])
--> 147 return [parse(s) for s in text.splitlines()]

File ~/projects/napalm/napalm/base/utils/string_parsers.py:147, in <listcomp>(.0)
    144 def parse(line: str) -> Tuple[str, ...]:
    145     return tuple([str(s.decode()) for s in unpack(line.ljust(len).encode())])
--> 147 return [parse(s) for s in text.splitlines()]

File ~/projects/napalm/napalm/base/utils/string_parsers.py:145, in parse_fixed_width.<locals>.parse(line)
    144 def parse(line: str) -> Tuple[str, ...]:
--> 145     return tuple([str(s.decode()) for s in unpack(line.ljust(len).encode())])

File ~/projects/napalm/napalm/base/utils/string_parsers.py:145, in <listcomp>(.0)
    144 def parse(line: str) -> Tuple[str, ...]:
--> 145     return tuple([str(s.decode()) for s in unpack(line.ljust(len).encode())])

UnicodeDecodeError: 'utf-8' codec can't decode bytes in position 8-9: unexpected end of data
```

Explicitly setting the encode/decode to ascii at least indicates that you really shouldn't base passing non-ASCII data into this function, making the exception clearer:

```python
In [2]: parse_fixed_width("12345689🧁1234567890123456", 10, 10, 10)
---------------------------------------------------------------------------
UnicodeEncodeError                        Traceback (most recent call last)
Cell In[2], line 1
----> 1 parse_fixed_width("12345689🧁1234567890123456", 10, 10, 10)

File ~/projects/napalm/napalm/base/utils/string_parsers.py:147, in parse_fixed_width(text, *fields)
    144 def parse(line: str) -> Tuple[str, ...]:
    145     return tuple([str(s.decode("ascii")) for s in unpack(line.ljust(len).encode("ascii"))])
--> 147 return [parse(s) for s in text.splitlines()]

File ~/projects/napalm/napalm/base/utils/string_parsers.py:147, in <listcomp>(.0)
    144 def parse(line: str) -> Tuple[str, ...]:
    145     return tuple([str(s.decode("ascii")) for s in unpack(line.ljust(len).encode("ascii"))])
--> 147 return [parse(s) for s in text.splitlines()]

File ~/projects/napalm/napalm/base/utils/string_parsers.py:145, in parse_fixed_width.<locals>.parse(line)
    144 def parse(line: str) -> Tuple[str, ...]:
--> 145     return tuple([str(s.decode("ascii")) for s in unpack(line.ljust(len).encode("ascii"))])

UnicodeEncodeError: 'ascii' codec can't encode character '\U0001f9c1' in position 8: ordinal not in range(128)
```